### ISSUE TITLE
HttpEngine QueryAsync Never Returns

### ISSUE THREAD

**Issue by Swiftless** [vivet/GoogleApi#128]
Hi,

I was previously using a much older version of the package on NuGet and upon updating to anything available on NuGet (3.6.7 onwards), none of my requests are ever completing.

As far as I'm aware the POST and GET requests are being made, my output is showing via Application Insights the URL and the method used.

I've appeared to be able to narrow it down to HttpEngine.cs -> QueryAsync being async. If I remove the async and inner await(s) from QueryAsync, everything runs fine. I am calling this through the Query call.

If I leave it as async, taskCompletion.SetResult appears to be called within the ContinueWith but never actually returns, leaving the thread stuck and the Task result never being available.

Thanks

**Comment by Swiftless** [comment]
Any request made to the "https://maps.googleapis.com/maps/api/place/textsearch/json?key=abc&query=123&language=en" Google API endpoint. I didn't try other parts of the library.

It may be a thread deadlocking issue, we are using your package inside of an ASP.NET project and it was working fine on previous versions. 

The quickest fix was as mentioned, to simply remove the async keyword on the QueryAsync function and remove the inner awaits. From what I can see the implementation of TaskCompletionSource appears correct so I'm not personally sure why it's not returning the task result.

**Comment by vivet** [comment]
Hi
Okay, the request you are creating isn't a problem, it seems simple and straight forward. To me it seems to be related to an async process outside the library, but within your application. Unless i have access to you application, or the part of the application where the problem resides, it's hard for me to help much. Though, first of all which .NET version are you using, and also try add ```.ConfigureAwait(false)``` from where you execute the query (or the call yo your own method async method wrapping the GoogleApi invocation).

**Comment by Swiftless** [comment]
Thanks for the super quick response, Unfortunately no, the issue is does not appear to be related to anywhere within my application but to ASP.NET WebForms projects in general if you don't have async functions going down from Page_Load to every function.

This is not feasible which is why I am using the .Query() call and not the .QueryAsync() call. However because .Query is calling .QueryAsync we are seeing the deadlocking since there's a mix of synchronous and asynchronous code. I've tried calling the .QueryAsync version directly but it results in the same deadlocking issue.

One example of converting everything to async not being feasible is async functions don't allow passing parameters by reference which is required in many parts of my code.

For now I've got a version of your code that drops the async keyword from QueryAsync and that appears to work but obviously isn't ideal for direct usage of QueryAsync by other developers.

It might be better to have the .Query version use a less async pathway for its inner logic to maintain compatibility.

**Comment by vivet** [comment]
Hi
Do you use .net framework? 
And which version?
Did you try using .ConfigureAwait(false) ?
Last question, do you use .Result somwhere?

**Comment by Swiftless** [comment]
I've tried .NET 4.6.1 and 4.7 and both fail with ASP.NET.

Yes I tried .ConfigureAwait(false) and it didn't help.

There is no need to use .Result as .Query() calls .Result within it and isn't an async function. However I have tried .Result on .QueryAsync() and it also fails to return unless I convert all my functions from Page_Load down to async, which as mentioned is not possible.

.Query can't rely on .QueryAsync().Result as it's causing deadlocking on threads and has broken compatibility with ASP.NET applications.

**Comment by vivet** [comment]
Ok
I read about this issue sometime ago, and if I remember correctly it's an issue with net framework. Nevertheless,I will help you solve it.

One more question are you using HttpClient elsewhere in the invocations leading to the GoogleApi call? I am asking cause the library re-uses the HttpClient, and therefore could be related.

**Comment by vivet** [comment]
This is the issue i am referring to 
http://blog.stephencleary.com/2012/07/dont-block-on-async-code.html

**Comment by vivet** [comment]
As I understand the ```.Result``` seems to cause the issue for you. 
Could you try change the Query method to use ```ConfigureAwait(false)``` and ```GetResult()``` and see how it behaves. 

Sorry I can't reproduce it myself.

**Comment by Swiftless** [comment]
Yes I have tried that.

If you create a brand new ASP.NET WebForms application and do the following in Page_Load, you can replicate the issue:

```
Dim request As New GoogleApi.Entities.Places.Search.Text.Request.PlacesTextSearchRequest With {
                .Key = "123",
                .Query = "abc"
            }

Dim response = GoogleApi.GooglePlaces.TextSearch.Query(request)
```
It will hang on the call to .Query because it deadlocks on the QueryAsync().Result call.

To fix the issue you have to change Page_Load to async (which it is not by default).

However it is not possible to convert all functions where GoogleApi is used to async because as mentioned, you lose the ability to use reference parameters which we require.

Because I require this in a synchronous context I am using .Query and not .QueryAsync on purpose. However because .Query simply calls .QueryAsync we receive deadlocking issues with a mix of async and sync code.

.Query (being the synchronous implementation) needs its inner logic reworked to not simply use .QueryAsync as the current implementation has broken compatibility.

**Comment by vivet** [comment]
Yes the inner workings of the Query method might needs its own implementation.

I will try to reproduce the issue

**Comment by vivet** [comment]
Hi
I have successfully reproduced the error, and I will investigate a solution.
I hope to have a package ready later today.

**Comment by vivet** [comment]
I just added ConfigureAwait(false) and it fixed it for me.
```
    protected void Page_Load(object sender, EventArgs e)
    {
        GoogleApi.GooglePlaces.TextSearch.QueryAsync(new PlacesTextSearchRequest
        {
            Key = "abc",
            Query = "abc"
        }).ConfigureAwait(false);
    }
``´

**Comment by vivet** [comment]
Sorry, nevermind my last message, it fails again when trying to get the actual resilt

**Comment by vivet** [comment]
Ok i have made a fix.
I can send you a dll so you can also confirm its working. Whats you email (just delete message afterwards, or send me an email first at vivet@vivetonline.com)

**Comment by vivet** [comment]
I deleted my last message.
I just published NuGet update, so should be there in 5-10 minutes. 

Please give the library a star if you like it :-)

**Comment by Swiftless** [comment]
Great job, that update has fixed it :)
